### PR TITLE
dbg pybind cpp lib / double diags exposed by dbg pybind cpp lib

### PR DESCRIPTION
**Pull request by PhilipDeegan**
includes https://github.com/PHAREHUB/PHARE/pull/517

if cmake is configured via CMAKE_BUILD_TYPE=Debug, or not configured with CMAKE_BUILD_TYPE (which defaults to Debug), two pybind libraries are built, one with existing functionality, and one with extra debug info, in this PR that includes diagnostic quantities being written in double precision, for double values.

in python, we check to see if we're in \_\_debug\_\_ ,  if not (python -O), we return the normal cpp lib, otherwise we attempt to load the debug library, falling back on the regular library if debug is not present.

### CODE REVIEW

File: src/python3/CMakeLists.txt [hunk 1]
```diff
@@ -11,3 +11,14 @@ set_target_properties(cpp
 )
 # this is on by default "pybind11_add_module" but can interfere with coverage so we disable it if coverage is enabled
 set_property(TARGET cpp PROPERTY INTERPROCEDURAL_OPTIMIZATION ${PHARE_INTERPROCEDURAL_OPTIMIZATION})
+
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+  pybind11_add_module(cpp_dbg cpp_simulator.cpp)
+  target_link_libraries(cpp_dbg PUBLIC phare_simulator)
+  target_compile_options(cpp_dbg PRIVATE ${PHARE_FLAGS} -DPHARE_HAS_HIGHFIVE=${PHARE_HAS_HIGHFIVE} -DPHARE_DIAG_DOUBLES=1 -DPHARE_CPP_MOD_NAME=cpp_dbg)
```
Comment (PhilipDeegan) [inline src/python3/CMakeLists.txt:18]: "-DPHARE_DIAG_DOUBLES=1" should probably be "-DPHARE_DEBUG_MODE=1"

and either "PHARE_DIAG_DOUBLES" is defined to true/1 if PHARE_DEBUG_MODE is active, or drop PHARE_DIAG_DOUBLES entirely and just use PHARE_DEBUG_MODE to pick between float/double in h5writer.h "using FloatType = std::conditional_t<PHARE_DIAG_DOUBLES, double, float>;"

File: pyphare/pyphare/cpp/__init__.py [hunk 2]
```diff
@@ -1,17 +1,21 @@
 
 
+def cpp_lib():
+    import importlib
+    if not __debug__:
+        return importlib.import_module("pybindlibs.cpp")
+    try:
+        return importlib.import_module("pybindlibs.cpp_dbg")
+    except ImportError as err:
+        return importlib.import_module("pybindlibs.cpp")
```
Comment (nicolasaunai) [inline pyphare/pyphare/cpp/__init__.py:10]: if I understand correctly:, we could run in debug, but for some reason can't improt the debug cpp version then we'd "silently" fallback on the regular one? do we want that to be silent?

Comment (PhilipDeegan) [reply pyphare/pyphare/cpp/__init__.py:10]: python doesn't know what mode cmake was built with, this assumes if there's no dbg library, cmake wasn't built in debug

File: src/python3/CMakeLists.txt (hunk 1 above)
Comment (nicolasaunai) [reply src/python3/CMakeLists.txt:18]: I think we might, for some reason, want to have the option of using double diags beyond debug mode so I'd keep that option, and yes keep setting it to true when in debug also, is useful

File: pyphare/pyphare/cpp/__init__.py (hunk 2 above)
Comment (nicolasaunai) [reply pyphare/pyphare/cpp/__init__.py:10]: what's this variable then __debug__

Comment (nicolasaunai) [reply pyphare/pyphare/cpp/__init__.py:10]: oh that's the python -O probably?

File: src/python3/CMakeLists.txt (hunk 1 above)
Comment (PhilipDeegan) [reply src/python3/CMakeLists.txt:18]: "__debug__" is false when exectuting with "python -O script.py"

File: pyphare/pyphare/cpp/__init__.py (hunk 2 above)
Comment (PhilipDeegan) [reply pyphare/pyphare/cpp/__init__.py:10]: "\_\_debug\_\_" is false when exectuting with "python -O script.py"

**Pull request closed without merging**